src/itensor.jl

```diff
@@ -695,3 +696,44 @@ function Base.read(io::IO,::Type{ITensor};kwargs...)
   end
   return T
 end
+
+
+"""
+    real_if_close(x ; tol = 1e-10) 
+
+Returns the real part of `x` if `x` is nearly real, i.e. `(x |> imag |> abs) < tol`; errors otherwise.
+
+This function differs from numpy's real_if_close in two ways in that it errors, rather than returning a complex result, when the argument is not nearly real.
+
+This choice is in part for type stability, and in part because when one uses a function like this one typically has good physical reason to believe that the answer should be real. Failure to in fact be nearly real will be indicative of a deeper problem, and likely to cause further problems downstream.
+"""
+real_if_close(x :: Real; tol = 1e-10) = x
+
+function real_if_close(x :: Complex{T}; tol = 1e-10) :: T where T <: Real 
+  if abs( imag(x) ) < tol
+    return real(x)
+  else
+    error("real_if_close: imaginary part of $x too large")
+  end
+end
+
+# turns the tensor into an array,
+# applies real_if_close elementwise,
+# and turns the result back into a tensor
+#
+# not sure this is really a good way to do this
+# in particular it seems bad for non-dense storage types
+# but this is a super useful function to have
+# (e.g. if you have a Hermitian MP(D)O
+# that you've put into a basis of Hermitian onsite operators:
+# it should be be at least gauge-equivalent to an MP(D)O with real tensors,
+# and typically IME it already has real tensors
+# ---but only up to 1e-13 or something)
+# (CDW)
+"""
+    real_if_close(A :: ITensor; tol = 1e-10)
+
+Return the real part of an itensor `A` if all the elements of A are nearly real, i.e. `all((x .|> imag .|> abs) < tol)`; errors otherwise.
+
+"""
+real_if_close(A :: ITensor; tol = 1e-10) = ITensor( real_if_close.(A |> array, tol=tol)  , inds(A))
```

Comment:
In general we don't want to convert to an Array, since that will be bad for sparse storage types (my vision for the `array` function was that it turned an ITensor into it's closest dense array version, so for Diag and BlockSparse it would fill with zeros. I call it `array` to specify that if the ITensor is `Dense` already, the storage won't get copied, which is a common Julia convention).

I can see the point of this function, since you might want to take the real part only under certain conditions. Is there a "better" norm for determining if the complex part should be thrown away, as opposed to an element-wise check? What about `norm(imag(A)) < 1e-10`?

Reply:
Also, your current implementation points to the fact that we should have broadcasting defined for ITensors, which is definitely something we plan to do. Then you could just do `real_if_close.(A,tol=tol)` for the elementwise version.

Reply:
Actually, I forgot that I already defined broadcast on the Tensor type, so you may be able to just do:
```julia
ITensor(real_if_close.(tensor(A), tol=tol))
```
(assuming we end up going with the elementwise version).

Reply:
Aha! This is much better. This seems like a decent, compact idiom; honestly, I'd be happy to just let the user do this any time she needs the functionality.

Reply:
Also could we add a branch `isreal(A) && return A` so this is a no-op for real ITensors? If we don't have an `isreal` function for ITensors, we should add it.

Reply:
Good thinking; my impulse is to do it with `eltype(A) <: Real && return A`. (If the tensors were parametrized over type we could do it at compile time with

    real_if_close(A :: ITensor{<:Real, N}, tol) = A

. They're not, are they? I didn't just miss that?)

Reply:
They are not, so you will have to do it at runtime with a function (either `isreal(A)` or `eltype(A) isa Real`, but really we would just define `isreal(A) = eltype(A) isa Real`). 

The reason is that ITensors are supposed to be flexible enough to change their data type "in-place", which would not be possible if they are parametrized by the element type. This allows for more generic code that handles in-place operations more automatically. For example, `scale!(A,3+im)` just works for an ITensor, whether the ITensor has complex or real storage (in the case of real storage, a new complex storage is made and replaces the real storage). It adds a bit of runtime overhead, but as long as performance critical code is within a type-stable function (the function barrier technique), we have seen that the overhead is minimal. That is actually part of the goal of the `Tensor` design: `ITensor`s pass their storage to `Tensor` functions, which are then type stable and performant, so the only overhead is an extra runtime dispatch to choose which `Tensor` function to dispatch on.